xbmc/guilib/GUIFontTTF.cpp

```diff
@@ -455,16 +442,51 @@ void CGUIFontTTF::DrawTextInternal(CGraphicContext& context,
         if (!ch)
           continue;
 
-        textWidth += ch->m_advance;
+        const float nextWidth = textWidth + ch->m_advance;
 
-        if (textWidth > maxPixelWidth)
+        if (nextWidth > maxPixelWidth)
         {
           // Start rendering from the glyph that does not exceed the maximum width
           startPosGlyph = std::distance(itRGlyph, glyphs.crend());
           break;
         }
+        textWidth = nextWidth;
       }
     }
+    else
+    {
+      // Determines the maximum width the text can have within the maximum width
+      if (alignment & XBFONT_TRUNCATED)
+        textWidth = ellipsesWidth;
+
+      for (const Glyph& glyph : glyphs)
+      {
+        Character* ch = GetCharacter(text[glyph.m_glyphInfo.cluster], glyph.m_glyphInfo.codepoint);
+        if (!ch)
+          continue;
+
+        const float nextWidth = textWidth + ch->m_advance;
+
+        if (nextWidth > maxPixelWidth)
+          break;
+
+        textWidth = nextWidth;
+      }
+    }
+
+    if (alignment & XBFONT_RIGHT)
+    {
+      // Moves the x pos with the purpose of having the text effect aligned to the right
+      startX += maxPixelWidth - textWidth;
+    }
+
+    if (alignment & XBFONT_CENTER_X && !(alignment & XBFONT_RIGHT) && !(alignment & XBFONT_LEFT))
+    {
+      if (alignment & XBFONT_CENTER_X)
```

Comment:
Which case is the first if() meant to filter? There is overlap with the second if()

A validation that alignment doesn't contain conflicting flags could be good but would belong to the beginning of the function (or its caller), would have some error logging and would override the flags to go back to a normal situation for the rest of the code.

Reply:
i made an attempt and as i was thinking it print on log millions lines of validating errors

so i will add a kid of validation check in to `CGUIFontTTF::DrawTextInternal` inside the `if (dirtyCache)`,
so that when cached dont print anymore tons of log lines
maybe not perfect but at least users/devs can know if has set an invalid alignment that can produce wrong rendering

Reply:
added validation as said on my previous comment, please try give a look